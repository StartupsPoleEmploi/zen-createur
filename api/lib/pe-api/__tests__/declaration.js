/* eslint-disable no-await-in-loop */
const config = require('config');
const nock = require('nock');
const { sendDeclaration } = require('../declaration');
const { DECLARATION_STATUSES } = require('../../../constants');

const declarationWithEmployers = {
  id: 543,
  userId: 165,
  hasWorked: true,
  hasTrained: false,
  hasInternship: false,
  hasSickLeave: false,
  hasMaternityLeave: false,
  hasRetirement: false,
  hasInvalidity: false,
  isLookingForJob: true,
  jobSearchStopMotive: null,
  hasFinishedDeclaringEmployers: true,
  isFinished: false,
  createdAt: '2019-01-02T15:55:29.957Z',
  updatedAt: '2019-01-07T11:48:53.839Z',
  monthId: 8,
  isTransmitted: false,
  isEmailSent: false,
  isDocEmailSent: false,
  employers: [
    {
      employerName: 'Test',
      workHours: 23,
      salary: 34,
      hasEndedThisMonth: false,
      id: 1735,
      userId: 165,
      declarationId: 543,
    },
  ],
};

const declarationWithEmployersAndHighWorkHours = {
  ...declarationWithEmployers,
  employers: [
    {
      employerName: 'Test',
      workHours: 500, // should be sent as MAX_DECLARABLE_HOURS
      salary: 34,
      hasEndedThisMonth: false,
      id: 1735,
      userId: 165,
      declarationId: 543,
    },
  ],
};

const declarationWithoutEmployers = {
  hasWorked: false,
  hasTrained: false,
  hasInternship: false,
  hasSickLeave: false,
  hasMaternityLeave: false,
  hasRetirement: false,
  hasInvalidity: false,
  isLookingForJob: true,
  jobSearchStopMotive: null,
  ignoreErrors: false,
  userId: 165,
  monthId: 8,
  hasFinishedDeclaringEmployers: true,
  isTransmitted: false,
  isFinished: false,
};

const declarationWithAllDatesFilled = {
  id: 543,
  userId: 165,
  hasWorked: true,
  hasTrained: false,
  hasInternship: true,
  infos: [
    {
      type: 'internship',
      startDate: '2018-12-11T23:00:00.000Z',
      endDate: '2019-01-10T23:00:00.000Z',
    },
    {
      type: 'sickLeave',
      startDate: '2018-12-17T23:00:00.000Z',
      endDate: '2019-01-18T23:00:00.000Z',
    },
    {
      type: 'maternityLeave',
      startDate: '2018-12-23T23:00:00.000Z',
    },
    {
      type: 'retirement',
      startDate: '2018-12-18T23:00:00.000Z',
    },
    {
      type: 'invalidity',
      startDate: '2018-12-24T23:00:00.000Z',
    },
    {
      type: 'jobSearch',
      endDate: '2018-12-25T23:00:00.000Z',
    },
  ],
  hasSickLeave: true,
  hasMaternityLeave: true,
  hasRetirement: true,
  hasInvalidity: true,
  isLookingForJob: false,
  jobSearchStopMotive: 'work',
  hasFinishedDeclaringEmployers: true,
  isFinished: false,
  createdAt: '2019-01-02T15:55:29.957Z',
  updatedAt: '2019-01-07T11:54:01.296Z',
  monthId: 8,
  isTransmitted: false,
  isEmailSent: false,
  isDocEmailSent: false,
  employers: [
    {
      employerName: 'IUI',
      workHours: 232,
      salary: 41,
      hasEndedThisMonth: true,
      userId: 165,
      declarationId: 543,
    },
    {
      employerName: 'Test',
      workHours: 23,
      salary: 34,
      hasEndedThisMonth: false,
      id: 1735,
      userId: 165,
      declarationId: 543,
    },
  ],
};

const declarationWithJobSearchMotiveRetirement = {
  ...declarationWithAllDatesFilled,
  jobSearchStopMotive: 'retirement',
};
const declarationWithJobSearchMotiveOther = {
  ...declarationWithAllDatesFilled,
  jobSearchStopMotive: 'other',
};

const declarationWithMultipleSicknessAndInternshipDates = {
  ...declarationWithoutEmployers,
  hasSickLeave: true,
  hasInternship: true,
  infos: [
    {
      type: 'internship',
      startDate: '2018-12-01T23:00:00.000Z',
      endDate: '2018-12-08T23:00:00.000Z',
    },
    {
      type: 'internship',
      startDate: '2018-12-11T23:00:00.000Z',
      endDate: '2019-01-10T23:00:00.000Z',
    },
    {
      type: 'sickLeave',
      startDate: '2018-12-02T23:00:00.000Z',
      endDate: '2018-12-08T23:00:00.000Z',
    },
    {
      type: 'sickLeave',
      startDate: '2018-12-17T23:00:00.000Z',
      endDate: '2019-01-18T23:00:00.000Z',
    },
  ],
};

const accessToken = 'AZERTYUIOP';
const userId = 1;

describe('PE API: sendDeclaration', () => {
  describe('API call success', () => {
    it('should send formatted data for declarations', async () => {
      let parsedBody;
      let parsedHeaders;

      const declarations = [
        declarationWithEmployers,
        declarationWithEmployersAndHighWorkHours,
        declarationWithoutEmployers,
        declarationWithAllDatesFilled,
        declarationWithJobSearchMotiveRetirement,
        declarationWithJobSearchMotiveOther,
        declarationWithMultipleSicknessAndInternshipDates,
      ];

      const scope = nock(config.apiHost)
        .post(
          '/partenaire/peconnect-actualisation/v1/actualisation',
          (body) => {
            parsedBody = body;
            return body;
          },
        )
        .times(declarations.length)
        .reply(function reply() {
          parsedHeaders = this.req.headers;
          return [200, { statut: DECLARATION_STATUSES.SAVED }];
        });

      for (const declaration of declarations) {
        const {
          body: { statut },
        } = await sendDeclaration({
          declaration,
          accessToken,
          userId,
          // Arbitrary, just to set forceIncoherences to true once
          ignoreErrors: declaration.jobSearchStopMotive === 'other',
        });
        expect(statut).toEqual(DECLARATION_STATUSES.SAVED);
        expect(parsedHeaders).toMatchSnapshot();
        expect(parsedBody).toMatchSnapshot();
      }
      scope.done();
    });
  });

  describe('API call failure', () => {
    it('should fail if there was a server error', (done) => {
      const scope = nock(config.apiHost)
        .post('/partenaire/peconnect-actualisation/v1/actualisation')
        .reply(500, {
          statut: DECLARATION_STATUSES.TECH_ERROR,
          message: 'Everything just exploded',
        });
      sendDeclaration({
        declaration: declarationWithEmployers,
        accessToken,
        userId,
      }).catch((err) => {
        expect(err).toBeDefined();
        scope.done();
        done();
      });
    });

    it('should retry multiple times if there was a status "impossible or unavailable" as an answer', (done) => {
      const scope = nock(config.apiHost)
        .post('/partenaire/peconnect-actualisation/v1/actualisation')
        .times(3)
        .reply(200, {
          statut: DECLARATION_STATUSES.IMPOSSIBLE_OR_UNNECESSARY,
          message: 'Actualisation non effectuÃ©e',
        });
      sendDeclaration({
        declaration: declarationWithEmployers,
        accessToken,
        userId,
      }).then(({ body }) => {
        expect(body.statut === DECLARATION_STATUSES.IMPOSSIBLE_OR_UNNECESSARY);
        scope.done();
        done();
      });
    });
  });
});
