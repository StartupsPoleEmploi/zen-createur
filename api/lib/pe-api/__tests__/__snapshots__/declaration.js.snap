// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 1`] = `
Object {
  "accept": "application/json",
  "accept-encoding": "gzip",
  "authorization": "Bearer AZERTYUIOP",
  "content-length": 57,
  "content-type": "application/json",
  "host": "api-r.es-qvr.fr",
  "media": "I",
  "user-agent": "node-superagent/3.8.3",
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 2`] = `
Object {
  "forceIncoherence": 0,
  "montSalaire": 34,
  "nbHeuresTrav": 23,
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 3`] = `
Object {
  "accept": "application/json",
  "accept-encoding": "gzip",
  "authorization": "Bearer AZERTYUIOP",
  "content-length": 58,
  "content-type": "application/json",
  "host": "api-r.es-qvr.fr",
  "media": "I",
  "user-agent": "node-superagent/3.8.3",
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 4`] = `
Object {
  "forceIncoherence": 0,
  "montSalaire": 34,
  "nbHeuresTrav": 420,
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 5`] = `
Object {
  "accept": "application/json",
  "accept-encoding": "gzip",
  "authorization": "Bearer AZERTYUIOP",
  "content-length": 22,
  "content-type": "application/json",
  "host": "api-r.es-qvr.fr",
  "media": "I",
  "user-agent": "node-superagent/3.8.3",
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 6`] = `
Object {
  "forceIncoherence": 0,
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 7`] = `
Object {
  "accept": "application/json",
  "accept-encoding": "gzip",
  "authorization": "Bearer AZERTYUIOP",
  "content-length": 298,
  "content-type": "application/json",
  "host": "api-r.es-qvr.fr",
  "media": "I",
  "user-agent": "node-superagent/3.8.3",
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 8`] = `
Object {
  "dateDebutMaladie": "18122018",
  "dateDebutMaternite": "24122018",
  "dateDebutStage": "12122018",
  "dateFinMaladie": "19012019",
  "dateFinRech": "26122018",
  "dateFinStage": "11012019",
  "dateInvalidite": "25122018",
  "dateRetraite": "19122018",
  "forceIncoherence": 0,
  "montSalaire": 75,
  "motifFinRech": 0,
  "nbHeuresTrav": 255,
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 9`] = `
Object {
  "accept": "application/json",
  "accept-encoding": "gzip",
  "authorization": "Bearer AZERTYUIOP",
  "content-length": 298,
  "content-type": "application/json",
  "host": "api-r.es-qvr.fr",
  "media": "I",
  "user-agent": "node-superagent/3.8.3",
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 10`] = `
Object {
  "dateDebutMaladie": "18122018",
  "dateDebutMaternite": "24122018",
  "dateDebutStage": "12122018",
  "dateFinMaladie": "19012019",
  "dateFinRech": "26122018",
  "dateFinStage": "11012019",
  "dateInvalidite": "25122018",
  "dateRetraite": "19122018",
  "forceIncoherence": 0,
  "montSalaire": 75,
  "motifFinRech": 1,
  "nbHeuresTrav": 255,
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 11`] = `
Object {
  "accept": "application/json",
  "accept-encoding": "gzip",
  "authorization": "Bearer AZERTYUIOP",
  "content-length": 298,
  "content-type": "application/json",
  "host": "api-r.es-qvr.fr",
  "media": "I",
  "user-agent": "node-superagent/3.8.3",
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 12`] = `
Object {
  "dateDebutMaladie": "18122018",
  "dateDebutMaternite": "24122018",
  "dateDebutStage": "12122018",
  "dateFinMaladie": "19012019",
  "dateFinRech": "26122018",
  "dateFinStage": "11012019",
  "dateInvalidite": "25122018",
  "dateRetraite": "19122018",
  "forceIncoherence": 1,
  "montSalaire": 75,
  "motifFinRech": 2,
  "nbHeuresTrav": 255,
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 13`] = `
Object {
  "accept": "application/json",
  "accept-encoding": "gzip",
  "authorization": "Bearer AZERTYUIOP",
  "content-length": 134,
  "content-type": "application/json",
  "host": "api-r.es-qvr.fr",
  "media": "I",
  "user-agent": "node-superagent/3.8.3",
}
`;

exports[`PE API: sendDeclaration API call success should send formatted data for declarations 14`] = `
Object {
  "dateDebutMaladie": "03122018",
  "dateDebutStage": "02122018",
  "dateFinMaladie": "19012019",
  "dateFinStage": "11012019",
  "forceIncoherence": 0,
}
`;
